### ISSUE TITLE
support for X509 authentication if ResourceServer is enabled

### ISSUE THREAD

**Issue by Captain-P-Goldfish** [spring-projects/spring-security-oauth#1363]
Hi, I am struggling with the spring security configuration that gets broken down if I use the **@EnableResourceServer**  annotation. Doing so breaks several things in my application:

1. I also use the **@EnableGlobalMethodSecurity(prePostEnabled = true)** annotation and if I secure my methods with the **@PreAuthorize("hasRole('ROLE_ADMIN')")** annotation I get:

    `org.springframework.security.authentication.AuthenticationCredentialsNotFoundException: An Authentication object was not found in the SecurityContext
	at org.springframework.security.access.intercept.AbstractSecurityInterceptor.credentialsNotFound(AbstractSecurityInterceptor.java:379)
	at org.springframework.security.access.intercept.AbstractSecurityInterceptor.beforeInvocation(AbstractSecurityInterceptor.java:223)
	at org.springframework.security.access.intercept.aopalliance.MethodSecurityInterceptor.invoke(MethodSecurityInterceptor.java:65)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:689)
	at my.package$$EnhancerBySpringCGLIB$$15ec1c04.getAll(<generated>)`

    for `SecurityContextHolder.getContext().getAuthentication()` is null

2. I tried to secure the very same method as before with client-authentication based on the detail that the necessary role gets only granted if ssl-client-authentication was performed. This is also broken down:

     The `OAuth2AuthenticationProcessingFilter` gets added after the `X509AuthenticationFilter` resulting      in cleared context with the execution of `OAuth2AuthenticationProcessingFilter` and therefore with no authentication for the user, eventhough the authentication has been performed successfully.

is there any way to work around this mess?

**Comment by jgrandja** [comment]
@Captain-P-Goldfish It's not a common use case to enable `HttpSecurity.x509()` and `@EnableResourceServer` in the same configuration.

`@EnableResourceServer` is used on OAuth 2.0 Resource Servers that protect resources and allow/authorize access based on the contents (authorities) of the incoming access token.

Whereas, client certificate authentication is used to authenticate a user via the X509 certificate.

It doesn't make sense to me why you would enable both. Can you please provide me more details around your requirements so I can better understand?

**Comment by Captain-P-Goldfish** [comment]
We have setup a normal OAuth2 system. But to configure this system we are using restful service in the same application. To provide high security on the configuration endpoints we want to use authentication via X509 certificates that will also allow us to track any change on the configuration by knowing exactly who accessed these endpoints.  (of course only as long as the keys are not corrupted.)

**Comment by jgrandja** [comment]
> we want to use authentication via X509 certificates that will also allow us to track any change on the configuration by knowing exactly who accessed these endpoints

If your requirement is as simple as tracking/logging the user who is accessing the protected resource than you can simply extract this information from the access token if it's stored in the access token. For example, a JWT can store the `sub` claim or some custom claim `user-id` which the Resource Server can extract and than use to record the user accessing it - or whatever you choose to do with it. 

You would need to configure the Authorization Server to include the user identifier claim when it mints the access token. This is a typical pattern that I would recommend over using x509 auth.

**Comment by Captain-P-Goldfish** [comment]
this is not our goal. The configuration endpoints are only to be accessed by the provider admins. the OAuth2 endpoints are to be accessed by arbitary users. 
We just want to ensure that the configuration is not accessed by anyone who might know/guess a password of an admin.

Besides we have a pretty special workflow. The users that authorize the clients to access their resources are not registered with us. It is more like this that they authorize us to provide highly reliable informations to the client that we get for a very short period during the authentication process. The client is then able to access the user data knowing that the data has been verified and is reliable.

So in other words we have actually 2 different workflows in our application. The user OAuth2 workflow that is usable for anyone.
And endpoints to configure the application that should only be accessed by admin. The SSL client authentication here is just to ensure that no unauthorized personnel is messing with the database and to track any change that was made on the database.

**Comment by jgrandja** [comment]
> The configuration endpoints are only to be accessed by the provider admins.

You also have the option to protect the configuration endpoints using different scopes compared to the scopes that allow access to the user-specific OAuth2 endpoints.

For example, you could do this for your Admin endpoints using `@EnableResourceServer`:

```
	@Override
	public void configure(HttpSecurity http) throws Exception {
		http
			.mvcMatcher("/admin/configuration/**")
			.authorizeRequests()
				.mvcMatchers("/admin/configuration/**").access("#oauth2.hasAnyScope('admin.config.read', 'admin.config.write')");
	}
```

The admin application would go through the `authorization_code` flow and request access to the scopes `admin.config.read` and `admin.config.write`. The admin user would grant/authorize those scopes and than the client/application would be able to access the admin endpoints.

The access token that is granted to the client for the user-specific OAuth2 endpoints should not have these scopes attached and therefore won't be able to access which is your primary goal.

**Comment by Captain-P-Goldfish** [comment]
In our case that is not quite what we want. We would have to implement a totally different Oauth2 workflow. Our authorization process requires additional devices and identity cards. Meaning that we hold user-data only within a cache and we will never have them registered. Doing what you suggest would mean implementing a toatlly different additional workflow which is quite some work again.

I created a second tomcat connector now that demands client authentication and the unsecured connector is blocked by accessing the configuration endpoints with an additional filter. This is far less work than setting up another database tables for admin users etc. 
Would be really great if something like this would be taken into account when implementing OpenID Connect in spring-security 5.

**Comment by jgrandja** [comment]
> Our authorization process requires additional devices and identity cards

You never did mention that in the earlier comment and yes that would require some extra work.

> Would be really great if something like this would be taken into account when implementing OpenID Connect in spring-security 5

Feel free to log a new feature request in Spring Security 5 and as much detail as possible.

Also, seems like you have come up with your solution so please close this issue if you are good on your end.
